pkg/storage/azure/azure_test.go

```diff
@@ -329,6 +333,138 @@ func TestConfigEnvWithUserKey(t *testing.T) {
+	resp := &http.Response{
+		StatusCode: http.StatusOK,
+		Request:    r,
+		Body:       io.NopCloser(bytes.NewBufferString(s.body)),
+	}
+	s.response = append(s.response, resp)
+	return resp, nil
+}
+
+func TestUserProvidedTags(t *testing.T) {
+	for _, tt := range []struct {
+		name         string
+		userTags     []configv1.AzureResourceTag
+		expectedTags map[string]string
+		infraName    string
+		responseBody string
+	}{
+		{
+			name:      "no-user-tags",
+			infraName: "some-infra",
+			// only default tags
+			expectedTags: map[string]string{
+				"kubernetes.io_cluster.some-infra": "owned",
+			},
+			responseBody: `{"nameAvailable":true}`,
+		},
+		{
+			name:      "with-user-tags",
+			infraName: "test-infra",
+			userTags: []configv1.AzureResourceTag{
+				{
+					Key:   "tag1",
+					Value: "value1",
+				},
+				{
+					Key:   "tag2",
+					Value: "value2",
+				},
+			},
+			// default tags and user tags
+			expectedTags: map[string]string{
+				"kubernetes.io_cluster.test-infra": "owned",
+				"tag1":                             "value1",
+				"tag2":                             "value2",
+			},
+			responseBody: `{"nameAvailable":true}`,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			sender := &sender{
+				body: tt.responseBody,
+			}
+
+			storageConfig := &imageregistryv1.ImageRegistryConfigStorageAzure{}
+
+			drv := NewDriver(context.Background(), storageConfig, nil)
+			drv.authorizer = autorest.NullAuthorizer{}
+			drv.sender = sender
+
+			_, _, err := drv.assureStorageAccount(
+				&Azure{
+					SubscriptionID: "subscription-id",
+					ResourceGroup:  "resource-group",
+				},
+				&configv1.Infrastructure{
+					Status: configv1.InfrastructureStatus{
+						InfrastructureName: tt.infraName,
+						Platform:           configv1.AzurePlatformType,
+						PlatformStatus: &configv1.PlatformStatus{
+							Type: configv1.AzurePlatformType,
+							Azure: &configv1.AzurePlatformStatus{
+								ResourceTags: tt.userTags,
+							},
+						},
+					},
+				},
+			)
+			if err != nil {
+				t.Errorf("unexpected error %q", err)
+			}
+
+			// flag to confirm presence of tags
+			foundTags := false
+
+			for _, resp := range sender.response {
+				if resp != nil && resp.Request != nil && resp.Request.Body != nil {
+
+					reqBody := make(map[string]interface{})
+					if err := json.NewDecoder(resp.Request.Body).Decode(&reqBody); err != nil {
+						t.Fatalf("error decoding request: %q", err)
+					}
+
+					// ignore request without tags
+					if _, ok := reqBody["tags"]; ok {
+						foundTags = true
+						if tags, ok := reqBody["tags"].(map[string]interface{}); ok {
+
+							// convert into correct type
+							receivedTags := make(map[string]string)
+							for k, v := range tags {
+								receivedTags[k] = fmt.Sprintf("%+v", v)
+							}
+
+							// compare the tags
+							if !reflect.DeepEqual(tt.expectedTags, receivedTags) {
+								t.Fatalf(
+									"expected tags %+v, received tags %+v",
```

Comment:
There's a handy lib called `cmp` that will output diffs between two values. Since this isn't a simple string I think it's worth using it to output the diff here.
Example: https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/storage/s3/s3_test.go#L158-L160

Reply:
oh, nice. I was not aware of this lib. Sure, this can be changed to 

```
t.Fatalf(
		"unexpected tags: %s",
		cmp.Diff(tt.expectedTags, receivedTags),
	)
```